Bring in 5.1 - 5.0 abi compatibility

matthewloring opened the pull request.

ofrobots commented:
Good work! Looking at the diff from v8.h in 5.0, there is still this ABI difference remaining:

``` diff
@@ -7149,7 +7323,7 @@ class Internals {
       1 * kApiPointerSize + kApiIntSize;
   static const int kStringResourceOffset = 3 * kApiPointerSize;

-  static const int kOddballKindOffset = 4 * kApiPointerSize;
+  static const int kOddballKindOffset = 5 * kApiPointerSize;
   static const int kForeignAddressOffset = kApiPointerSize;
```

matthewloring commented:
That should be fixed now.

ofrobots commented:
Looking at some of the other include/ files, a few more differences. These should be easy to fix however:

``` patch
--- a/deps/v8/include/v8-debug.h
+++ b/deps/v8/include/v8-debug.h
@@ -18,13 +18,11 @@ enum DebugEvent {
   Exception = 2,
   NewFunction = 3,
   BeforeCompile = 4,
-  AfterCompile  = 5,
+  AfterCompile = 5,
   CompileError = 6,
-  PromiseEvent = 7,
-  AsyncTaskEvent = 8,
+  AsyncTaskEvent = 7,
 };
--- a/deps/v8/include/v8-platform.h
+++ b/deps/v8/include/v8-platform.h
@@ -152,9 +152,9 @@ class Platform {
    */
   virtual uint64_t AddTraceEvent(
       char phase, const uint8_t* category_enabled_flag, const char* name,
-      uint64_t id, uint64_t bind_id, int32_t num_args, const char** arg_names,
-      const uint8_t* arg_types, const uint64_t* arg_values,
-      unsigned int flags) {
+      const char* scope, uint64_t id, uint64_t bind_id, int32_t num_args,
+      const char** arg_names, const uint8_t* arg_types,
+      const uint64_t* arg_values, unsigned int flags) {
     return 0;
   }
```

matthewloring commented:
These should now be compatible.

verwaest-zz commented on deps/v8/src/objects.h:
```diff
@@ -708,12 +708,12 @@ enum InstanceType {
-  JS_VALUE_TYPE,               // LAST_CUSTOM_ELEMENTS_RECEIVER
   JS_MESSAGE_OBJECT_TYPE,
   JS_DATE_TYPE,
-  JS_OBJECT_TYPE,
```
Swapping this back around will be bad for performance, since we perform range checks on LAST_SPECIAL_RECEIVER_TYPE and LAST_CUSTOM_ELEMENTS_RECEIVER. If JS_OBJECT_TYPE falls above, all JS_OBJECT_TYPE objects (most common JavaScript object) will take the slow path for e.g., .hasOwnProperty.

matthewloring commented:
I added custom checks for `JS_OBJECT_TYPE` at the comparison points to `LAST_CUSTOM_ELEMENTS_RECEIVER`. Using this contrived benchmark:

``` javascript
var o = {};
for (var i = 0; i < 10000000; i++) {
  Object.assign(o, { i: i });
  for (var k in o) {
    if (!o.hasOwnProperty(k)) {
      console.log('fail');
    }
  }
}
```

I observed:

```
mattloring-macbookpro:node mattloring$ time node-6.1.0 ~/Desktop/slow-test.js

real    0m1.985s
user    0m1.961s
sys 0m0.019s
mattloring-macbookpro:node mattloring$ time out/Release/node-old ~/Desktop/slow-test.js

real    0m25.168s
user    0m24.408s
sys 0m0.095s
mattloring-macbookpro:node mattloring$ time out/Release/node-fast ~/Desktop/slow-test.js

real    0m2.104s
user    0m1.960s
sys 0m0.016s
```

@ofrobots @verwaest  What do you think?

ofrobots commented on deps/v8/src/builtins.cc:
```diff
@@ -481,9 +481,13 @@ void Builtins::Generate_ObjectHasOwnProperty(
                                              LAST_CUSTOM_ELEMENTS_RECEIVER)),
+                          assembler->Word32NotEqual(
+                              instance_type, assembler->Int32Constant(
+                                             JS_OBJECT_TYPE))),
```
I think this is not right. JS_SPECIAL_API_OBJECT_TYPE used to be <= LAST_CUSTOM_ELEMENTS_RECEIVER before your change, but is > after your change. JS_OBJECT_TYPE is invariant here.

I think the expression you want is:

```
OR (
     LESS_EQUAL (instance_type, LAST_CUSTOM_ELEMENTS_RECEIVER
     EQUAL(instance_type, JS_SPECIAL_API_OBJECT_TYPE))
```

matthewloring replied:
Fixed.

ofrobots commented:
LGTM to me. Let's get @verwaest to review as well.

targos commented:
There is now https://github.com/v8/v8/commit/ea5e96ff05035e5d3fc22ce4df0a9b401c0fae5a in the 5.1 branch. Do you have to make changes for that?

verwaest-zz commented:
LGTM as well.

matthewloring commented:
I've updated the patch to support https://github.com/v8/v8/commit/ea5e96ff05035e5d3fc22ce4df0a9b401c0fae5a.

ofrobots commented:
LGTM. @jeisinger PTAL as well.

jeisinger commented:
LGTM

targos commented:
@matthewloring If everything is ready, could you squash the commits so I can pick them up in https://github.com/nodejs/node/pull/7016 ?

matthewloring commented:
@targos All squashed.

targos commented:
The patch does not apply cleanly on [my branch](https://github.com/targos/node/tree/v8-5.1)

ofrobots commented:
Actually this PR needs more work. Intercepted enumerator tests are failing for me.

@matthewloring I think the problem is that `JS_SPECIAL_API_OBJECT_TYPE` has been reordered wrt. `LAST_CUSTOM_ELEMENTS_RECEIVER`. API objects (with interceptors) no longer look like they require custom elements iteration breaking enumerator tests.

@targos: this is not likely to be ready before next week.

ofrobots commented on deps/v8/include/v8-debug.h:
```diff
@@ -20,7 +20,8 @@ enum DebugEvent {
   AfterCompile = 5,
   CompileError = 6,
-  AsyncTaskEvent = 7,
+  PromiseEvent = 7,
```
Looking at this more closely, I think the removal of `PromiseEvent` was actually an API change that is missing from the [official list](https://docs.google.com/document/d/1g8JFi8T_oAE_7uAri7Njtig7fKaPDfotU6huOa1alds/edit?hl=en&forcehl=1) (/cc @natorion). I think we will have bring this back as a revert of https://codereview.chromium.org/1833563002.

Commit:
deps: revert removal of V8::PromiseEvent

The removal of the promise debug event is an API/ABI breaking change.

https://codereview.chromium.org/1833563002

ofrobots commented on deps/v8/src/builtins.cc:
```diff
@@ -223,7 +223,8 @@ inline bool PrototypeHasNoElements(Isolate* isolate, JSObject* object) {
     Map* map = prototype->map();
-    if (map->instance_type() <= LAST_CUSTOM_ELEMENTS_RECEIVER) return false;
+    if (map->instance_type() <= LAST_CUSTOM_ELEMENTS_RECEIVER ||
+      map->instance_type() == JS_GLOBAL_PROXY_TYPE) return false;
```
nit: align with the previous line.

Commit:
Bring in 5.1 - 5.0 abi compatibility

ofrobots commented:
LGTM from my side. Given that that the roots list was reordered; it would be good to get signoff from @verwaest or @jeisinger as well.

jeisinger commented:
lgtm

ofrobots commented:
Launched a CI: https://ci.nodejs.org/job/node-test-commit/3684/
@targos this is rebased on top of your branch.

matthewloring commented:
Closing this as it has landed in core.
